13_sandboxes/test_case_generator.py

```diff
@@ -0,0 +1,274 @@
+        "cd password-analyzer && poetry install --no-root && poetry run pytest",
+    )
+    print(p.stdout.read())
+    print(p.stderr.read())
+
+    # Run the generated test file
+    p = sb.exec(
+        "bash",
+        "-c",
+        f"cp /data/outputs/{file_name} password-analyzer/tests/ && cd password-analyzer && poetry run pytest",
+    )
+    print(p.stdout.read())
+    print(p.stderr.read())
+
+    sb.terminate()
+
+
+@app.function(volumes={"/data": files_volume})
+def main():
+    deepseek = Deepseek()
+    input_files = download_files_to_volume.remote(
+        folder_paths=["src/password_strength", "tests"]
+    )
+
+    print("Generating tests for: ", input_files)
+    output_files = deepseek.generate.map(input_files)
+    for output_file in output_files:
+        print(output_file)
+        run_sandbox(output_file)
+
+
+# # Addenda
+# The below functions are utility functions.
+
+
+def get_user_prompt(file_text: str, test_file_text: str) -> str:
+    return f"""
+    You are an expert Python test engineer. Your task is to improve an existing test file using `pytest`.
+
+    Step-by-step:
+    1. Carefully read the existing test file (below) and understand the current test cases.
+    2. Then read the source file (also below) and understand the function behavior, focusing on docstrings, edge cases, and argument types.
+    3. Based on that understanding, **add** new test cases to the test file to increase coverage—especially edge cases, boundary conditions, and untested branches.
+    4. Use `pytest` idioms, but do **not** add or change import statements—**use only what is already imported**.
+    5. Do **not** explain your reasoning—just return the final modified test file.
+
+    ### Requirements:
+    - Your output must be a valid, complete Python file with the added test cases.
+    - Do not modify existing test logic unless necessary to support your new test cases.
+    - Do not import any additional modules.
+    - This file will be run directly using `pytest`, so it must be immediately runnable.
+
+    --- BEGIN TEST FILE ---
+    {test_file_text}
+    --- END TEST FILE ---
+
+    --- BEGIN SOURCE FILE ---
+    {file_text}
+    --- END SOURCE FILE ---
+    """
+
+
+def get_system_prompt():
+    return (
+        "You are a senior software engineer with expertise in test-driven development and Python unit testing. "
+        "You write clean, idiomatic unit tests using the `pytest` framework, based on source code with functions or classes. "
+        "Your task is to enhance an existing test file by adding more test cases. Focus on edge cases, input validation, and untested behavior, especially as inferred from docstrings and type hints. "
+        "Do not change or add import statements. Do not explain your reasoning. Output only a complete, valid Python file. "
+        "Limit each line to a maximum of 100 characters to avoid output truncation or formatting errors."
+    )
+
+
+def post_process(output: str) -> str:
```

Comment:
Would be preferable to use structured outputs/function calling/JSONSchema here -- which is I think much more work coming from `transformers` and probably a reason to use SGLang or vLLM.